File: erpnext/accounts/report/royalty_payment_report/royalty_payment_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Royalty Payment Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": ("Company"),
			"fieldtype": "Link",
 			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
            "fieldname": "cost_center",
            "label": __("Parent Cost Center"),
			"fieldtype": "Link",
            "width": "80",
			"options": "Cost Center",
			get_query: () => {
				var company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						disabled: 0,
						company: company,
						is_group: 1
					}
				};
			},
		},
		{
			"fieldname": "branch",
			"label": ("Branch"),
			"fieldtype": "Link",
			"options": "Cost Center",
			get_query: () => {
				var cost_center = frappe.query_report.get_filter_value("cost_center")
				var company = frappe.query_report.get_filter_value("company")
				if(cost_center!= 'Natural Resource Development Corporation Ltd - NRDCL')
				{
					return {
						filters: {
							disabled: 0,
							company: company,
							parent_cost_center: cost_center
						}
					};
				}
				else
				{
					return {
						filters: {
							disabled: 0,
							company: company,
							is_group: 0
						}
					};
				}
			}
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_start_date")
		},
		{
			"fieldname": "to_date",
			"label": ("To Date"),
			"fieldtype": "Date",
			"default": frappe.defaults.get_user_default("year_end_date")
		},
		{
			"fieldtype":"Break"
		},
		{
			"fieldname": "workflow_state",
			"label": ("Workflow State"),
			"fieldtype": "Select",
			"options":["","Approved","Draft","Rejected", "Rejected by Supervisor", "Waiting Approval", "Waiting Supervisor Approval"]
		},
		{
			"fieldname": "payment_state",
			"label": ("Payment State"),
			"fieldtype": "Select",
			"options":["","Paid","Due","None"]
		},
		{
			"fieldname": "detail",
			"label": ("Show in Detail"),
			"fieldtype": "Check",
			"default": 0
		},
		{
			"fieldname": "range",
			"label": ("Range"),
			"fieldtype": "Link",
			"options": "Range"
		},
		{
			"fieldname": "item_sub_group",
			"label": ("Material Sub Group"),
			"fieldtype": "Link",
			"options": "Item Sub Group"
		},
		{
			"fieldtype":"Break"
		},
		{
			"fieldname": "timber_class",
			"label": ("Timber Class"),
			"fieldtype": "Link",
			"options": "Timber Class"
		},
		{
			"fieldname": "ref_id",
			"label": ("Royalty Ref ID"),
			"fieldtype": "Link",
			"options": "Royalty Payment"
		},
		{
			"fieldname": "payment_id",
			"label": ("Payment Ref ID"),
			"fieldtype": "Link",
			"options": "Journal Entry"
		},
		{
			"fieldname": "quantity",
			"label": ("Quantity"),
			"fieldtype": "Float"
		},
		{
			"fieldname": "rate",
			"label": ("Royalty Rate"),
			"fieldtype": "Float"
		}
		


	]
}

File: erpnext/accounts/report/royalty_payment_report/royalty_payment_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from erpnext.accounts.utils import get_child_cost_centers

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_data(filters):
	cond = ""
	payment_cond = ""
	if filters.get("from_date") and filters.get("to_date"):
		cond += " and rp.posting_date between '{0}' and '{1}'".format(filters.get("from_date"), filters.get("to_date"))
	if filters.get("workflow_state"):
		cond += " and rp.workflow_state = '{0}'".format(filters.get("workflow_state"))
	if filters.get("cost_center"):
		if not filters.get("branch"):
			all_ccs = get_child_cost_centers(filters.get("cost_center"))
			cond += " and rp.branch in (select b.name from `tabBranch` b where b.cost_center in {0})".format(tuple(all_ccs))
		else:
			branch_name = frappe.db.get_value("Branch",{"cost_center": filters.get("branch")}, "name") 
			cond += " and rp.branch = '{0}'".format(branch_name)
	if filters.get("payment_state"):
		payment_cond += " and data.status = '{0}'".format(filters.get("payment_state"))
	
	if filters.get("range"):
		cond += " and rp.range_name = '{0}'".format(filters.get("range"))

	if filters.get("ref_id"):
		cond += " and rp.name = '{0}'".format(filters.get("ref_id"))
	
	if filters.get("payment_id"):
		cond += " and rp.journal_entry = '{0}'".format(filters.get("payment_id"))
	
	if filters.get("item_sub_group"):
		if filters.get("detail"):
			if filters.get("item_sub_group") in ('Firewood, Bhutan Furniture','BBPL firewood','Firewood(BBPL)','Firewood (Bhutan Ply)','Firewood','Post','Bakals','Woodchips','Briquette','Off-cuts/Sawn timber waste','Off-Cuts','Saw Dust'):
				cond += " and rpa.particular in (select item_name from `tabItem` where item_sub_group = '{0}' )".format(filters.get("item_sub_group"))
			else:
				cond += " and rpa.particular = '{0}'".format(filters.get("item_sub_group"))
		else:
			frappe.throw("This filter in only applicable in detailed mode")
	
	if filters.get("timber_class"):
		if filters.get("detail"):
			cond += " and rpa.timber_class = '{0}'".format(filters.get("timber_class"))
		else:
			frappe.throw("This filter in only applicable in detailed mode")
	
	if filters.get("quantity"):
		if filters.get("detail"):
			cond += " and rpa.quantity = '{0}'".format(filters.get("quantity"))
		else:
			frappe.throw("This filter in only applicable in detailed mode")
	
	if filters.get("rate"):
		if filters.get("detail"):
			cond += " and rpa.rate = '{0}'".format(filters.get("rate"))
		else:
			frappe.throw("This filter in only applicable in detailed mode")

	# query = frappe.db.sql("""
	# 	select *from ( select ml.branch, ml.production_type, ml.posting_date, ml.name
    #             from
    #             `tabMarking List` ml  
    #             where ml.docstatus =1 {0}) as m
    #             left join
    #             (select rp.name, rp.journal_entry, rp.total_royalty, rp.marking_list
    #             from
	# 	`tabRoyalty Payment` rp 
    #             where rp.docstatus =1) as r
    #             on m.name = r.marking_list""".format(cond))
	
	#Above query commented and below query added for new Royalty Payment Report //Kinley Dorji 2021-02-17
	if not filters.get("detail"):
		query = frappe.db.sql("""
			select * from (
			select rp.posting_date as posting_date, rp.name as name, rp.workflow_state as workflow_state, rp.branch as branch, rp.range_name as range_name, rp.production_type as production_type, ifnull(rp.journal_entry,'None') as journal_entry, rp.business_activity as business_activity, rp.total_royalty, rp.discount_amount, rp.net_royalty, rp.total_qty, rp.less_qty, rp.net_qty,
			CASE WHEN (select je.docstatus from `tabJournal Entry` je where je.name = rp.journal_entry and rp.docstatus = 1) = 1 THEN 'Paid'
			WHEN (select je.docstatus from `tabJournal Entry` je where je.name = rp.journal_entry and rp.docstatus = 1) = 0 THEN 'Due'
			ELSE 'None' END AS status	
			from `tabRoyalty Payment` rp where rp.docstatus != 2 {0}
			) as data where 1 = 1 {1}
		""".format(cond, payment_cond))
	else:
		# if filters.get()
		query = frappe.db.sql("""
			select * from (
			select rp.posting_date as posting_date, rpa.particular , rp.name as name, rp.workflow_state as workflow_state, rp.branch as branch, rp.range_name as range_name, rp.production_type as production_type, ifnull(rp.journal_entry,'None') as journal_entry, rp.business_activity as business_activity, rpa.quantity, rpa.rate, rpa.amount, rpa.timber_class, rpa.reading,
			CASE WHEN (select je.docstatus from `tabJournal Entry` je where je.name = rp.journal_entry and rp.docstatus = 1) = 1 THEN 'Paid'
			WHEN (select je.docstatus from `tabJournal Entry` je where je.name = rp.journal_entry and rp.docstatus = 1) = 0 THEN 'Due'
			ELSE 'None' END AS status	
			from `tabRoyalty Payment` rp, `tabRoyalty Payment Adhoc` rpa where rp.name = rpa.parent and rp.docstatus != 2 {0}
			) as data where 1 = 1 {1}
		""".format(cond, payment_cond))
	return query

def get_columns(filters):
	if not filters.get("detail"):
		columns = [
			("Date") + "::80",
			("Royalty Payment ID") + ":Link/Royalty Payment:150",
			("Workflow State") + "::100",	
			("Branch") + ":Link/Branch:150",
			("Range Name") + "::150",
			("Production Type") + "::140",
			# ("Marking List ID") + ":Link/Marking List:120",
			("Journal Entry") + ":Link/Journal Entry:120",
			("Business Activity") + "::140",
			("Total Royalty") + ":Currency:140",
			("Discount") + ":Currency:140",
			("Net Royalty") + ":Currency:140",
			("Total Quantity") + ":Float:140",
			("Subracted Quantity") + ":Float:140",
			("Net Quantity") + ":Float:140",
			("Status") + "::140",
		]
	else:
		columns = [
			("Date") + "::80",
			("Particular") + "::150",
			("Royalty Payment ID") + ":Link/Royalty Payment:150",
			("Workflow State") + "::100",	
			("Branch") + ":Link/Branch:150",
			("Range Name") + "::150",
			("Production Type") + "::140",
			# ("Marking List ID") + ":Link/Marking List:120",
			("Journal Entry") + ":Link/Journal Entry:120",
			("Business Activity") + "::140",
			("Quantity") + ":Float:140",
			("Royalty Rate") + "::100",
			("Amount") + ":Float:140",
			("Timber Class") + "::100",
			("Reading") + "::100",
			("Status") + "::140",
		]
	return columns
